modules/weaver-tests/src/smithy4s/weavertests/WeaverHttpRequestTestCase.scala

```diff
@@ -0,0 +1,249 @@
+  * 4. Assertion are performed on the recorded input (from decoding the http request)
+  *
+  */
+class WeaverHttpRequestTestCase[
+    Alg[_[_, _, _, _, _]],
+    Op[_, _, _, _, _]
+](
+    client: (
+        HttpApp[IO],
+        Uri
+    ) => Either[UnsupportedProtocolError, smithy4s.Monadic[Alg, IO]],
+    server: smithy4s.Monadic[Alg, IO] => Either[
+      UnsupportedProtocolError,
+      HttpRoutes[IO]
+    ]
+)(implicit service: Service[Alg, Op], ce: CompatEffect)
+    extends Expectations.Helpers {
+  import ce._
+  import org.http4s.implicits._
+  private val baseUri = uri"http://localhost/"
+
+  private def matchRequest(
+      request: Request[IO],
+      testCase: HttpRequestTestCase
+  )(implicit loc: SourceLocation) = {
+    val bodyAssert = testCase.body
+      .map { expectedBody =>
+        request.bodyText.compile.string.map { responseBody =>
+          assert.eql(expectedBody, responseBody)
+        }
+      }
+      .getOrElse(success.pure[IO])
+
+    val headerAssert =
+      testCase.headers
+        .map { expectedHeaders =>
+          expectedHeaders
+            .map { case (name, value) =>
+              val actual = request.headers.get(CIString(name))
+              actual
+                .map { values =>
+                  val actualValue = values.map(_.value).fold
+                  assert.eql(value, actualValue)
+                }
+                .getOrElse(
+                  failure(s"Header $name was not found in the response.")
+                )
+            }
+            .reduce(_ && _)
+        }
+        .getOrElse(success)
+
+    val expectedUri = baseUri
+      .withPath(
+        Uri.Path.unsafeFromString(testCase.uri)
+      )
+      .withQueryParams(
+        testCase.queryParams.combineAll.map {
+          _.split("=", 2) match {
+            case Array(k, v) =>
+              (
+                k,
+                Uri.decode(
+                  toDecode = v,
+                  charset = StandardCharsets.UTF_8,
+                  plusIsSpace = true
+                )
+              )
+          }
+        }.toMap
+      )
+
+    val uriAssert = assert.eql(expectedUri, request.uri)
+    val methodAssert = assert.eql(
+      testCase.method.toLowerCase(),
+      request.method.name.toLowerCase()
+    )
+
+    List(
+      bodyAssert,
+      headerAssert.pure[IO],
+      uriAssert.pure[IO],
+      methodAssert.pure[IO]
+    ).sequence.map(_.reduce(_ && _))
+  }
+
+  private[weavertests] def makeServerTest[I, E, O, SE, SO](
+      endpoint: Endpoint[Op, I, E, O, SE, SO],
+      testCase: HttpRequestTestCase,
+      inputFromDocument: Document.Decoder[I]
+  ): GeneratedTest = {
+    GeneratedTest(
+      name = endpoint.id.toString + "(server|request): " + testCase.id,
+      assertions = {
+        val dummyOutput = DefaultSchemaVisitor(endpoint.output)
+        val input = testCase.params
+          .map { inputFromDocument.decode(_).liftTo[IO] }
+          .getOrElse(IO.unit.asInstanceOf[IO[I]])
```

Comment:
Can you do the cast on `Unit` level instead of `IO`?

Reply:
alternatively we could decode a `Document.obj()` with `inputFromDocument` normally.